autopilot/contractor.go

```diff
@@ -562,14 +570,39 @@ func (c *contractor) runContractChecks(ctx context.Context, w Worker, contracts
 		host.PriceTable.HostBlockHeight = state.cs.BlockHeight
 
 		// decide whether the host is still good
-		usable, unusableResult := isUsableHost(state.cfg, state.rs, gc, host.Host, minScore, revision.Filesize)
+		usable, unusableResult := isUsableHost(state.cfg, state.rs, gc, host.Host, minScore, contract.FileSize())
 		if !usable {
 			reasons := unusableResult.reasons()
 			toStopUsing[fcid] = strings.Join(reasons, ",")
 			c.logger.Infow("unusable host", "hk", hk, "fcid", fcid, "reasons", reasons)
 			continue
 		}
 
+		// if we were not able to get a valid price table for the host, but we
+		// did pass the host checks, we only want to be lenient if this contract
+		// is in the current set and only for a certain number of times,
+		// controlled by maxKeepLeeway
+		if invalidPT {
+			if _, found := inCurrentSet[fcid]; !found || remainingKeepLeeway == 0 {
+				toStopUsing[fcid] = "no valid price table"
+				continue
+			}
+			remainingKeepLeeway-- // we let it slide
+		}
+
+		// if we were not able to the contract's revision, we can't properly
+		// perform the checks that follow, however we do want to be lenient if
+		// this contract is in the current set and we still have leeway left
+		if contract.Revision == nil {
+			if _, found := inCurrentSet[fcid]; !found || remainingKeepLeeway == 0 {
+				toStopUsing[fcid] = errContractNoRevision.Error()
+			} else {
+				toKeep = append(toKeep, fcid)
+				remainingKeepLeeway-- // we let it slide
+			}
+			continue // can't perform contract checks without revision
+		}
```

Comment:
I think those two `if` need to be merged. e.g.

```go
if invalidPT || contract.Revision == nil {
  if _, found := inCurrentSet[fcid]; !found || remainingKeepLeeway == 0 {
    if invalidPT {
      toStopUsing[fcid] = "no valid price table"
    } else if contract.Revision == nil {
      toStopUsing[fcid] = errContractNoRevision.Error()
    }
    continue // can't perform further checks
  } 
  remainingKeepLeeway-- // we let it slide
}
```

Because otherwise we might subtract 2 from `remainingKeepLeeway` for a single contract if I'm reading this correctly.